### ISSUE TITLE
Mismatching number of %-placeholders in string formatting

### ISSUE THREAD

**Issue by kaddkaka** [astral-sh/ruff#24031]
### Summary

Incorrect number of format specifies should give error. 

There are already many good warnings and tips about using f-strings. Howerever this code does not give any warnings today:

```
var = f"{banana}" % banana
```

**Comment by ntBre** [comment]
Huh, I would have expected [percent-format-positional-count-mismatch (F507)](https://docs.astral.sh/ruff/rules/percent-format-positional-count-mismatch/#percent-format-positional-count-mismatch-f507) to fire here, but it does not in the playground, even without the f-string, which was my first guess at why it might be skipped. It looks like we only check tuple right-hand sides:

https://github.com/astral-sh/ruff/blob/471fd12c3991b69166f95b476affee3b1f048b88/crates/ruff_linter/src/rules/pyflakes/rules/strings.rs#L742-L744

Although the f-string then causes problems too: https://play.ruff.rs/89d0e453-1296-4b32-86b8-c5c15d114c9e

We should double check if there's a good reason for these restrictions in the git history, but otherwise this does seem like a false negative to me.

**Comment by ai-mountain** [comment]
I looked into this — F507 only fires when the right-hand side is a tuple, so a single value like `"hello" % banana` is skipped entirely. On top of that, the left-hand side check only matches plain string literals, so f-strings are never evaluated as %-format templates.

I think extending F507 to handle non-tuple right-hand sides and also flagging f-strings used with `%` would fix this. I'll check the git history for why these restrictions exist. Can I make a PR for this?

**Comment by ai-mountain** [comment]
I checked the git history — `List` and `Set` were intentionally removed from F507 in #5986 because `'%s' % [1, 2, 3]` is valid Python (only tuples get unpacked in `%` formatting). So that restriction is correct.

The gap is for non-tuple single values like `"hello %s %s" % banana` — F507 currently ignores these entirely. Since a non-tuple RHS always counts as exactly 1 positional argument, we could flag cases where the format string expects more than 1.

**Comment by MichaReiser** [comment]
I think the reason why single elements are skipped entirely is:


```pycon
>>> banana = (1, 2)
>>> "hello %s %s" % banana
'hello 1 2'
```

**Comment by ai-mountain** [comment]
Thanks @MichaReiser 
I will make another PR that fixes false negative for non-tuple LHS in %-formatting.
Should I open a new issue or just only PR for this?

**Comment by MichaReiser** [comment]
>  I will make another PR that fixes false negative for non-tuple LHS in %-formatting.

Can you give an example?

**Comment by ai-mountain** [comment]
I fixed the **RHS** case — when the right-hand side is a non-tuple value.

**Before my fix:**
```python
'%s %s' % 42        # No F507 — ruff only checked tuples on RHS
'%s %s' % "hello"   # No F507
'%s %s' % (1 + 2)   # No F507
```

**After my fix:**
```python
'%s %s' % 42        # F507 — now flagged (int is not a tuple)
'%s %s' % "hello"   # F507 — now flagged (str is not a tuple)
'%s %s' % (1 + 2)   # F507 — now flagged (int + int → int, not a tuple)
```

I will fix the **LHS** issue (f-string like `f"hello %s" % banana`)

**Comment by kaddkaka** [comment]
I think this means that this issue is not completely fixed. There are two more categories of cases to fix:

1. f-string on LHS
2. string without any placeholder but still using a `%`

The original example `f"{banana}" % banana` covers both these cases.

**Comment by kaddkaka** [comment]
Please reopen this issue I can't seem to do it @MichaReiser

**Comment by ai-mountain** [comment]
@kaddkaka  Could you open another issue?
I will make a new PR soon.

**Comment by MichaReiser** [comment]
I think we should open separate issues. I'm not convinced we should support `%` formatting within f-strings. Instead, we may want a rule that warns about using `%` in f-strings (unless there's a use case that I'm not seeing where using `%` over the f-string replacement is desired). But we should definitely discuss what we want to do here.

Using `%` with a string without any placeholders seems reasonable, but also something we can cover and track in a separate issue.

**Comment by ai-mountain** [comment]
Agreed, I am thinking about a separate PR for that — a new rule (RUF072) that warns about using `%` on an f-string, since mixing f-string interpolation with `%`-formatting is almost certainly a mistake. I also can't think of a valid use case for it.

I'll open a separate issue for the "string without any placeholders + `%`" case as well.
